politeia_sync.go

```diff
@@ -0,0 +1,476 @@
+		return errors.New(ErrListenerAlreadyExist)
+	}
+
+	p.notificationListeners[uniqueIdentifier] = notificationListener
+	return nil
+}
+
+func (p *Politeia) RemoveNotificationListener(uniqueIdentifier string) {
+	p.notificationListenersMu.Lock()
+	defer p.notificationListenersMu.Unlock()
+
+	delete(p.notificationListeners, uniqueIdentifier)
+}
+
+func (p *Politeia) publishNewProposal(proposal *Proposal) {
+	if p.readConfigFunc(PoliteiaNotificationConfigKey, false) {
+		p.notificationListenersMu.Lock()
+		defer p.notificationListenersMu.Unlock()
+
+		for _, notificationListener := range p.notificationListeners {
+			notificationListener.OnNewProposal(proposal.ID, proposal.CensorshipRecord.Token)
+		}
+	}
+}
+
+func (p *Politeia) publishVoteStarted(proposal *Proposal) {
+	if p.readConfigFunc(PoliteiaNotificationConfigKey, false) {
+		p.notificationListenersMu.Lock()
+		defer p.notificationListenersMu.Unlock()
+
+		for _, notificationListener := range p.notificationListeners {
+			notificationListener.OnProposalVoteStarted(proposal.ID, proposal.CensorshipRecord.Token)
+		}
+	}
+}
+
+func (p *Politeia) publishVoteFinished(proposal *Proposal) {
+	if p.readConfigFunc(PoliteiaNotificationConfigKey, false) {
+		p.notificationListenersMu.Lock()
+		defer p.notificationListenersMu.Unlock()
+
+		for _, notificationListener := range p.notificationListeners {
+			notificationListener.OnProposalVoteFinished(proposal.ID, proposal.CensorshipRecord.Token)
+		}
+	}
+}
+
+func (p *Politeia) IsSyncProgressListenerRegisteredFor(uniqueIdentifier string) bool {
+	p.syncData.mu.RLock()
+	_, exists := p.syncData.syncProgressListeners[uniqueIdentifier]
+	p.syncData.mu.RUnlock()
+	return exists
+}
+
+func (p *Politeia) AddSyncPropogressListener(syncProgressListener ProposalSyncProgressListener, uniqueIdentifier string) error {
+	if p.IsSyncProgressListenerRegisteredFor(uniqueIdentifier) {
+		return errors.New(ErrListenerAlreadyExist)
+	}
+
+	p.syncData.mu.Lock()
+	p.syncData.syncProgressListeners[uniqueIdentifier] = syncProgressListener
+	p.syncData.mu.Unlock()
+
+	// If sync is already on, notify this newly added listener of the current progress report.
+	return p.PublishLastSyncProgress(uniqueIdentifier)
+}
+
+func (p *Politeia) RemoveSyncProgressListener(uniqueIdentifier string) {
+	p.syncData.mu.Lock()
+	delete(p.syncData.syncProgressListeners, uniqueIdentifier)
+	p.syncData.mu.Unlock()
+}
+
+func (p *Politeia) PublishLastSyncProgress(uniqueIdentifier string) error {
+	p.syncData.mu.RLock()
+	defer p.syncData.mu.RUnlock()
+
+	syncProgressListener, exists := p.syncData.syncProgressListeners[uniqueIdentifier]
+	if !exists {
+		return errors.New(ErrInvalid)
+	}
+
+	if p.syncData.syncing {
+		switch p.syncData.syncStage {
+		case ProposalsDiscoverySyncState:
+			syncProgressListener.OnProposalsDiscovery()
+		case ProposalsFetchSyncState:
+			syncProgressListener.OnProposalsFetched(&p.syncData.proposalsFetchProgress)
+		}
+	}
+
+	return nil
+}
```

Comment:
What this sync does is fetch data from an API, a progress listener is not needed for this and it's not also implemented by the frontend.